images/cilium/Dockerfile

```diff
@@ -0,0 +1,45 @@
+# syntax=docker/dockerfile:1.1-experimental
+
+# Copyright 2020 Authors of Cilium
+# SPDX-License-Identifier: Apache-2.0
+
+ARG CILIUM_BUILDER_IMAGE=docker.io/cilium/cilium-builder-dev:5c1d8916b5ad0fbb86aabcc6239e163b6c645a71
+ARG CILIUM_RUNTIME_IMAGE=docker.io/cilium/cilium-runtime-dev:bf5a13ab0dd341b446ec1f40a809d365024eeb09
+
+FROM --platform=linux/amd64 ${CILIUM_BUILDER_IMAGE} as builder
+
+ARG NOSTRIP
+ARG LOCKDEBUG
+
+RUN --mount=type=bind,readwrite,target=/go/src/github.com/cilium/cilium --mount=target=/root/.cache,type=cache --mount=target=/go/pkg/mod,type=cache \
+  make clean-container build-container install-container \
+    NOSTRIP=$NOSTRIP LOCKDEBUG=$LOCKDEBUG PKG_BUILD=1 SKIP_DOCS=true DESTDIR=/out/linux/amd64
+
+RUN --mount=type=bind,readwrite,target=/go/src/github.com/cilium/cilium --mount=target=/root/.cache,type=cache --mount=target=/go/pkg/mod,type=cache \
+  env GOARCH=arm64 CC=aarch64-linux-gnu-gcc \
+    make clean-container build-container install-container \
+      NOSTRIP=$NOSTRIP LOCKDEBUG=$LOCKDEBUG PKG_BUILD=1 SKIP_DOCS=true DESTDIR=/out/linux/arm64 \
+      HOST_CC=aarch64-linux-gnu-gcc HOST_STRIP=aarch64-linux-gnu-strip
```

Comment:
I don't understand why we are duplicating these lines. To me this looks that this could be :
```
ARG GOARCH
ARG ARCH
ARG OS
RUN --mount=type=bind,readwrite,target=/go/src/github.com/cilium/cilium --mount=target=/root/.cache,type=cache --mount=target=/go/pkg/mod,type=cache \
  env GOARCH=$GOARCH CC=$ARCH-$OS-gnu-gcc \
    make clean-container build-container install-container \
      NOSTRIP=$NOSTRIP LOCKDEBUG=$LOCKDEBUG PKG_BUILD=1 SKIP_DOCS=true DESTDIR=/out/$OS/$GOARCH \
      HOST_CC=$ARCH-$OS-gnu-gcc HOST_STRIP=$ARCH-$OS-gnu-strip
```

which works for both `arm64` and `amd64` so we don't have to waste resources building 2 architectures if, in the end we will only copy one of them (line 38 `COPY --from=builder /out/${TARGETPLATFORM} /`)

Reply:
>  in the end we will only copy one of them 

The way to understand it is that there is an implicit loop and `TARGETPLATFORM` is the loop variable.

The build runs only for `amd64`, as it has `FROM --platform=linux/amd64` (line 9).
But there is no `--platfrom` flag on line 34, so that stages runs on each of the platforms.

Reply:
Also, why are we mounting `/go/pkg/mod`?

Reply:
> > in the end we will only copy one of them
> 
> The way to understand it is that there is an implicit loop and `TARGETPLATFORM` is the loop variable.
> 

But from line 9 to line 33 we are building for both archs at the same time, no?

Reply:
> But from line 9 to line 33 we are building for both archs at the same time, no?

No, that's from line 34 onwards.

> Also, why are we mounting `/go/pkg/mod`? Cilium vendors all code in the `vendor` directory

That is just a matter of re-using pattern, and do not that directory is mounted as cache, so if something does end-up there it will get cached, but if nothing - nothing will get cached.